lib/utils/spreadwork/spreadwork.go

```diff
@@ -0,0 +1,111 @@
+ * it under the terms of the GNU Affero General Public License as published by
+ * the Free Software Foundation, either version 3 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package spreadwork
+
+import (
+	"context"
+	"time"
+
+	"github.com/gravitational/trace"
+	"github.com/jonboulle/clockwork"
+)
+
+// ApplyOverTimeConfig is the configuration values for the ApplyOverTime method.
+type ApplyOverTimeConfig struct {
+	// MaxDuration is the max duration for the operation to complete.
+	// It does not account for time spent in API calls.
+	MaxDuration time.Duration
+
+	// BatchInterval is the time spent between API calls to prevent server pressure.
+	// Defaults to 1 second.
+	BatchInterval time.Duration
+
+	// MinBatchSize is the min batch size per iteration.
+	// Defaults to 5 items.
+	MinBatchSize int
+
+	// clock is used for tests only.
+	clock clockwork.Clock
+}
+
+// CheckAndSetDefaults checks that the MaxDuration was set.
+func (c *ApplyOverTimeConfig) CheckAndSetDefaults() error {
+	if c.MaxDuration == 0 {
+		return trace.BadParameter("max duration is required")
+	}
+
+	if c.BatchInterval == 0 {
+		c.BatchInterval = time.Second
+	}
+
+	if c.MaxDuration < c.BatchInterval {
+		return trace.BadParameter("max interval must be greater than batch interval")
+	}
+
+	if c.MinBatchSize == 0 {
+		c.MinBatchSize = 5
+	}
+
+	if c.clock == nil {
+		c.clock = clockwork.NewRealClock()
+	}
+
+	return nil
+}
+
+// ApplyOverTime applies the applyFn to all elements of items during a period of time.
+func ApplyOverTime[T any](ctx context.Context, conf ApplyOverTimeConfig, items []T, applyFn func(T)) error {
+	if err := conf.CheckAndSetDefaults(); err != nil {
+		return trace.Wrap(err)
+	}
+
+	maxBatches := int(conf.MaxDuration / conf.BatchInterval)
+	dynamicBatchSize := (len(items) / maxBatches) + 1
+	if dynamicBatchSize < conf.MinBatchSize {
+		dynamicBatchSize = conf.MinBatchSize
+	}
+
+	for {
+		if dynamicBatchSize > len(items) {
+			dynamicBatchSize = len(items)
+		}
+		chunk := items[0:dynamicBatchSize:dynamicBatchSize]
+		items = items[dynamicBatchSize:]
+
+		for _, c := range chunk {
+			// If ctx is Done, return without processing the other elements of the current chunk.
+			select {
+			case <-ctx.Done():
+				return nil
+			default:
+				applyFn(c)
+			}
+		}
+		if len(items) == 0 {
+			return nil
+		}
+
+		select {
+		// If ctx is Done, return without processing the remaining chunks.
+		case <-ctx.Done():
+			return nil
+		case <-conf.clock.After(conf.BatchInterval):
```

Comment:
I would create a ticker and "defer ticker.Stop()" it explicitly. I think Go has done some work recently to automatically garbage collect tickers but just in case.

Reply:
Changed to a ticker.

> If the apply function above takes longer than one tick it would result in immediately unblocking this instead of actually spreading out the work at a constant interval.

We define a Max Duration so that we don't exceed a given amount of time.
Having to wait exactly BatchInterval duration means we would most likely exceed that Max Duration value.

We can exceed it anyway, because we are never guaranteed on how long the `applyFn` is going to take.
But this way, the exceeding time should be much less.
If `applyFn` takes close to 0s, then it shouldn't matter whether we use a ticker or a timer.

This `spreadwork` can evolve to support other scenarios:
- start a goroutine that will process a given chunk at periodic intervals (one that wouldn't wait for the current chunk to be processed)
- always wait X amount of time between chunks (what you proposed)

VS what we currently have which does sequential processing and never starts the next chunk before `time.Now()+BatchInterval` (whether the BatchInterval is just a sleep or actual processing/`applyFn` time)